### PR TITLE
Yen's Algorithm

### LINKED ISSUE #613: Yen's Algorithm
I was doing a project that need the K-shortest loopless paths for a graph with non-negative edge cost, the Yen's Algorithm(https://en.wikipedia.org/wiki/Yen%27s_algorithm) was the best solution for me. I didn't found any impementation in Julia so I did it using LightGraphs. Do you have interest in adding the code to LightGraphs?

### PR DESCRIPTION
**Pull request by Thuener**
Implementation of the k-shortest loopless paths for a graph with non-negative edge cost using Yen's Algorithm(https://en.wikipedia.org/wiki/Yen%27s_algorithm) 

closes #613

### CODE REVIEW

File: src/shortestpaths/yen.jl [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    path = enumerate_paths(dj)[target]
+    if isempty(path)
+        return YenState{T, U}(Vector{T}(), Vector{Vector{U}}())
+    end
+
+    dists = Array{T,1}()
+    push!(dists,dj.dists[target])
+    A = [path]
+    B = PriorityQueue()
+    gc = deepcopy(g)
```
Comment (juliohm) [inline src/shortestpaths/yen.jl]: Perhaps rename this variable to something else? Julia defines the garbage collector as `gc` already.

Comment (Thuener) [inline src/shortestpaths/yen.jl]: Of course.

**Comment by juliohm** [comment]
Thanks for your work @Thuener, what do you think of passing the shortest path algorithm as an option instead of always using Dijkstra? This would give us more flexibility.

**Comment by Thuener** [comment]
There is a problem with definition in this case, Yen's algorithm uses Dijkstra.

**Comment by juliohm** [comment]
The wiki page says it is blind to the algorithm and that one can pick any
shortest path algo.

File: src/shortestpaths/yen.jl [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+Designed for yen k-shortest-paths calculations.
+"""
+struct YenState{T, U<:Integer}<: AbstractPathState
+    dists::Vector{T}
+    paths::Vector{Vector{U}}
+end
+
+
+"""
+  yen_k_shortest_paths(g, source, target, distmx=DefaultDistance(), K);
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: This should be indented 4 spaces, and the default value for `K` should be in the signature.

File: src/shortestpaths/yen.jl [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  Perform [Yen's algorithm](http://en.wikipedia.org/wiki/Yen%27s_algorithm)
+  on a graph, computing k-shortest distances between `source` and `target` other vertices.
+  Return a [`YenState`](@ref) that contains distances and paths.
+
+  ### Optional Arguments
+  - `k=1`: The number of paths to find.
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: upper case `k`? also, you don't need this section if you've defined the default in the signature (see prev. comment). If you decide to keep it, `distmx` is also an optional argument.

File: src/shortestpaths/yen.jl [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+"""
+function yen_k_shortest_paths(
+    g::AbstractGraph,
+    source::U,
+    target::U,
+    distmx::AbstractMatrix{T}=DefaultDistance(),
+    K::Int=1) where T where U<:Integer
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: Might want to bound `T <: Real`.

**Comment by Thuener** [comment]
@juliohm you are right, seans Dijkstra is just a suggestion.

**Comment by juliohm** [comment]
Thuener, agora que vi que você é brasileiro! Bom saber que a comunidade está crescendo :blush:

**Comment by Thuener** [comment]
Sim, aqui na PUC-Rio estamos usando bastante para otimização, principalmente o JuMP. Eu já dei alguns cursos sobre Julia aqui.

**Comment by sbromberger** [comment]
This is SO CLOSE. I'll add a few more line comments for a last cleanup and then final review. Thanks for sticking with it, @Thuener !

File: src/shortestpaths/yen.jl [hunk 5]
```diff
@@ -0,0 +1,98 @@
+"""
+    struct YenState{T, U}
+Designed for yen k-shortest-paths calculations.
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: blank line between indented signature and description.

File: src/shortestpaths/yen.jl [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+"""
+struct YenState{T, U<:Integer}<: AbstractPathState
+    dists::Vector{T}
+    paths::Vector{Vector{U}}
+end
+
+
+"""
+    yen_k_shortest_paths(g, source, target, distmx=DefaultDistance(), K=1);
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: indent signature please.

File: src/shortestpaths/yen.jl [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            for (u, v) in edgesremoved
+                add_edge!(gcopy, u, v)
+            end
+        end
+        if !isempty(B)
+            push!(dists,DataStructures.peek(B)[2])
+            push!(A,dequeue!(B))
+        else
+            break
+        end
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: This if block should be rewritten to simplify:

```
isempty(B) && break
push!(dists,DataStructures.peek(B)[2])
push!(A,dequeue!(B))
```

File: src/shortestpaths/yen.jl [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                    v = ppath[j + 1]
+                    if has_edge(gcopy, u, v)
+                        rem_edge!(gcopy, u, v)
+                        push!(edgesremoved,(u, v))
+                    end
+                end
+            end
+
+            # Remove node of root path
+            for n = 1:(length(rootpath)-1)
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: spacing here, please:
```
for n = 1:(length(rootpath) - 1)
```
(there are a few places where spacing will improve readability. Basically, it's a good idea to follow every comma with a space, and surround operations with spaces (except for `=` in function signatures). All code about this point looks good (except for L57), but please check the rest of it.

File: src/shortestpaths/yen.jl [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    if source == target
+        return YenState{T, U}([U(0)], [[source]])
+    end
+
+    dj = dijkstra_shortest_paths(g, source, distmx)
+    path = enumerate_paths(dj)[target]
+    if isempty(path)
+        return YenState{T, U}(Vector{T}(), Vector{Vector{U}}())
+    end
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: short-circuit here:
```
isempty(path) && return YenState{T, U}(Vector{T}(), Vector{Vector{U}}())
```

File: src/shortestpaths/yen.jl [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+function yen_k_shortest_paths(
+    g::AbstractGraph,
+    source::U,
+    target::U,
+    distmx::AbstractMatrix{T}=DefaultDistance(),
+    K::Int=1) where T <: Real where U<:Integer
+
+    if source == target
+        return YenState{T, U}([U(0)], [[source]])
+    end
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: short-circuit here:
```
source == target && return YenState{T, U}([U(0)], [[source]])
```

File: src/shortestpaths/yen.jl [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+            djspur = dijkstra_shortest_paths(gcopy, spurnode, distmx)
+            spurpath = enumerate_paths(djspur)[target]
+            if !isempty(spurpath)
+                # Entire path is made up of the root path and spur path
+                pathtotal = [rootpath[1:end-1];spurpath]
+                distpath  = dj.dists[spurnode] + djspur.dists[target]
+                # Add the potential k-shortest path to the heap
+                if !haskey(B,pathtotal)
+                    enqueue!(B, pathtotal, distpath)
+                end
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: Comment only: even though we're short-circuiting above, we don't short-circuit when there's a mutation, so keeping `enqueue!` in the if block like this is perfectly correct.

File: src/shortestpaths/yen.jl [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+struct YenState{T, U<:Integer}<: AbstractPathState
+    dists::Vector{T}
+    paths::Vector{Vector{U}}
+end
+
+
+"""
+    yen_k_shortest_paths(g, source, target, distmx=DefaultDistance(), K=1);
+
+    Perform [Yen's algorithm](http://en.wikipedia.org/wiki/Yen%27s_algorithm)
```
Comment (sbromberger) [inline src/shortestpaths/yen.jl]: unindent description, please.

**Comment by juliohm** [comment]
@Thuener sorry we are a little picky about formatting, we are looking forward to merge your PR. Please let us know when you have some time to fix those small issues.

We can also leave that suggestion on algorithms other than Dijkstra out of the table for now, if we need it in the future, we should be able to extend it easily.

Thanks again.

**Comment by Thuener** [comment]
It's ok, I'm just too busy now. I will catch up on this soon.

**Comment by sbromberger** [comment]
@Thuener - no worries; thank you very much for your contribution here. Let us know when you're ready to resume.

**Comment by Thuener** [comment]
In the weekend I will code the suggestion of using others algorithms than Dijkstra. If there are any other problem let me know.

**Comment by sbromberger** [comment]
@Thuener do you mind if I go ahead and merge this now, to get the functionality in? You can always make another PR to add the other algorithms.

**Comment by Thuener** [comment]
Go ahead, I don't mind.

**Comment by sbromberger** [comment]
Wait, I think you may have misunderstood. I want to commit these changes now, and then when you're ready to do other work, have you open up a separate PR. Is that OK?

**Comment by juliohm** [comment]
@Thuener, porque você fechou o PR? Está basicamente pronto para ser incorporado, não entendi...

Avisa se estiver tudo ok que reviso as suas alterações no código e faço o merge.

**Comment by Thuener** [comment]
Sorry, wrong button. It's ok.

**Comment by sbromberger** [comment]
Thanks, @Thuener !

**Pull request merged**